src/terminal/adapter/adaptDispatch.cpp

```diff
@@ -455,15 +458,32 @@ bool AdaptDispatch::CursorSavePosition()
 // - True if handled successfully. False otherwise.
 bool AdaptDispatch::CursorRestorePosition()
 {
-    unsigned int const uiRow = _coordSavedCursor.Y;
-    unsigned int const uiCol = _coordSavedCursor.X;
+    auto& savedCursorState = _savedCursorState[_usingAltBuffer];
+
+    auto uiRow = savedCursorState.Row;
+    auto uiCol = savedCursorState.Column;
+
+    // If the origin mode is relative, and the scrolling region is set (the bottom is non-zero),
+    // we need to make sure the restored position is clamped within the margins.
+    if (savedCursorState.IsOriginModeRelative && _srScrollMargins.Bottom != 0)
+    {
+        // VT origin is at 1,1 so we need to add 1 to these margins.
+        uiRow = std::clamp(uiRow, _srScrollMargins.Top + 1u, _srScrollMargins.Bottom + 1u);
```

Comment:
If someone sets scroll margins in the alt buffer, then swaps back to the main buffer, is this position not clamped by those margins from the alt buffer?

Reply:
That's a very good question. This isn't a problem for us in the alt buffer, since the margins are associated with the buffer and will be restored when we return to main (that's not correct, mind you, but that's a separate issue - the alt buffer needs a bit of work). There are other ways in which the restored position can end up outside the margins, though, and it's vital that these are clamped otherwise the cursor can end up in an illegal state. From the spec:

> If the restored cursor position lies outside of the scrolling region, and the restored value of Origin Mode is set, the Active position will be moved to the closest position within the scrolling region. That is, if the restored cursor position is above the Top Margin, the Active position will be set to the Top Margin, and if the restored position is below the Bottom Margin, the Active position will be set to the Bottom Margin. [...] This reason for this recovery is to guarantee that the Active Position will always be within the bounds of the scrolling region, and will never have a negative value.

I've just checked XTerm and this is actually something they get wrong (at least in the version I have). For example, this sequence will land the cursor position above the top margin.
 
    printf "\e[?6h\e[?1049h\e[10;20r\e[?1049l"

Now try querying the position with this:

    printf "\e[6n"; read

Note the returned y value will be something like 65529, which is clearly not correct.

Reply:
Actually, on further investigation, it seems our code is a little more broken than I originally thought. We're tracking the margins in two separate places, and they can get out of sync when using the alt buffer. But that's not really got anything to do with the save/restore code so is probably best dealt with in a separate issue.

Reply:
Please file another one for that mismatch so we don't lose track.

Reply:
It is on my TODO list. It just might take some time for me to put together a proper bug report.